Wasm/Tests.lean

```diff
@@ -0,0 +1,79 @@
+import LSpec
+import Megaparsec.Errors.Bundle
+import Megaparsec.Parsec
+import Wasm.Engine
+import Wasm.Wast.AST
+import Wasm.Wast.Parser
+import Wasm.Bytes
+
+open LSpec
+open Megaparsec.Errors.Bundle
+open Megaparsec.Parsec
+open Wasm.Engine
+open Wasm.Wast.AST.Module
+open Wasm.Wast.Parser
+open Wasm.Bytes
+
+namespace Wasm.Tests
+
+inductive ExternalFailure (e : IO.Error) : Prop
+instance : Testable (ExternalFailure e) := .isFailure s!"External failure: {e}"
+
+open Megaparsec.Errors.Bundle in
+inductive ParseFailure (src : String) (e : ParseErrorBundle Char String Unit) : Prop
+instance : Testable (ParseFailure src e) := .isFailure s!"Parsing:\n{src}\n{e}"
+
+open IO.Process (run) in
+def w2b (x : String) :=
+  run { cmd := "./wasm-sandbox", args := #["wast2bytes", x] } |>.toBaseIO
+
+open IO.Process (run) in
+def run_main (x : String) :=
+  run { cmd := "./wasm-sandbox", args := #["run_main", x] } |>.toBaseIO
+
+def runModule (m : Module) : Except EngineErrors Int := do
+  let store := mkStore m
+  let s₀ := Stack.mk []
+  match fidByName store "main" with
+  | .none => Except.error EngineErrors.function_not_found
+  | .some fid =>
+    run store fid s₀ >>= fun (_, s₁) =>
```

Comment:
I see two issues with this function/approach in general.

1. `run` already typechecks the results, so typechecking it here is superfluous.
2. More importantly, this can only handle single-function results, and an `Int` result at that, which is a weird restriction to apply after we've already supported multi-result functions. I understand this is the most basic of usecases, but I think it should either be clearly noted as such (and extended later) or fixed right away – my preferred outcome. `BEq` should work on `StackEntry`, so we can check the entire stack against expected instead of just one `Int`.

Reply:
1. Yes, but I need to get an int out of there, which I can only do via matching. In a way `run` will check if the code makes sense, whereas this function will check if the results of the code that makes sense were wrapped in a "main" function that's exactly `() -> I32`.
2. No, it's absolutely not important. If you look at wasm-sandbox test suite, you'll see that I test a lot of weird stuff and less weird stuff by testing / extracting the stuff I want to test in "main" wrapper. As a matter of fact, If I could have forced "main" to have shape `() -> Bool`, I would have.

To reiterate the point:

Your comment is analogous to saying that `describe` function or `test` function can't take arbitrarily-typed functions as its arguments. Yeah, it can't, because it's the function that facilitates testing.

I understand that the counter-argument could be "yes, but we have two engines that should be able to just run stuff and run eq on the outputs", but please remember that since we're using wasm tools via an external tool, it would require to also patch the external tool to spit out something that we can parse as Stack! Which is a week worth of work for me, I think. And THEN... We won't be testing anything! Because now numerous bugs in the implementation of Stack-as-WAST or Stack-as-Lean-Expression will cause our tests themselves to be buggy.

Finally, I don't agree that this design decision isn't documented. In the C4 commit, I clearly state that the solution to the problem of the lack of a blessed way to test runtime compatibility is to allow to call functions "main" with signatures "() -> I32" and compare their outputs between lean and a ref. impl. wrapper.

Reply:
OK, I see that `wasm-sandbox` itself only prints the first element on the stack. I really don't like that (and the lack of documentation there), but OK.

I pretty heavily disagree with multiple (basically all) points in your response, but I understand it's irrelevant because we just want to get started. As such, IMO this PR is a good first step that MUST be extended in the future.

The only thing is, I don't care that you wrote a good C4 commit message (or I do, but I don't think it's enough here). Developers'/code readers' flow does not automatically include reading `git log`, especially if they, the poor souls, aren't overly familiar with the system, and so the same points should be copy-pasted to code comments to not make them wonder. I'll do it for you since I also don't really like the way test errors are handled for runtime compatibility, and I'd like to make them better before we merge.

Reply:
Fair enough!